src/target/spirv/codegen_spirv.cc

```diff
@@ -561,6 +630,55 @@ void CodeGenSPIRV::VisitStmt_(const ForNode* op) {
 
   // Loop head
   builder_->StartLabel(head_label);
+
+  // In a normal matmul, the update semantics c += a * b is implemented by load and store to the
```

Comment:
Exactly right. I really liked your explanation, so I'll update the comment here based on it.

Reply:
I assume the intention here is to scalarize each 16x16 fragment and represent it using a variable.  This transformation does not bring performance gain compared to the approach of using a vector of 16x16 fragments and load/store from it since shader compiler can easily does the same scalarization.   Also, if you want to unroll the loop,  the address computation in each unrolled iteration is not CSEed.  So you will see a larger code size with all these non-CSEed address computations.

Reply:
Looks like my claim "Cooperative matrices cannot be loaded from / stored to a buffer" was based on a wrong assumption. I'd still argue that my approach is simpler and cleaner than the load / store based approach, even if there is no other benefit. A con of my approach is that it differs from how we treat vector types - I don't have a good answer for the question "why cooperative matrices should get special handling" other than somewhat subjective opinions. One thing that's not clear to me: I believe the spec doesn't say how many bytes each matrix takes up, so I am not sure it is possible to allocate a buffer of and address them. Maybe we don't need to worry about such issues since memory is typed in SPIRV and size / offset are in terms of the number of elements. 

I'm happy to change the implementation to be based on conventional load / store based approach if people prefer that way. I'd love to get @Lunderberg's opinion on this point.

I should also remove the term "unroll" in the comment since nothing is really unrolled by codegen. For example here is a good 4k matmul TIR https://gist.github.com/masahi/dc7173d39f4b376884f49019cde24826 and the generated spv https://gist.github.com/masahi/3fd481c8037b1ed55876c89305ec02ef. The two representations mostly map one to one (no more unrolling done on spv).

Reply:
The Mei's approach in https://github.com/apache/tvm/pull/14817, being based on the WMMA intrinsics originally developed for CUDA, inherits both pros and cons of CUDA WMMA support in TVM. In particular, it allows the same schedule to be compiled for CUDA and VK, and since CUDA WMMA is supported by our auto tensorization infra, auto tensorization might also work out of the box for VK. That's a big advantage. cc @vinx13

Reply:
Being able to reuse existing schedule and auto tensorization for WMMA is a big advantage. I'm wondering if we can reuse wmma intrinsics since we can still use fragment index to identify matrix (if we want to scalarize it)

Reply:
Looking through, I like Mei's approach as well.  It avoids making usage restrictions (e.g. buffers must be accessed through coop matrix primitives or load/store, but not both), and maintains the mutability semantics of both spirv/tir by avoiding long-lived coop matrix values in the generated spir-v.

Reply:
Ok thanks for the discussion, I'm closing my PR in favor of #14817 cc @mei-ye